### PR TITLE
Added `Exponential Moving Sum` to `UpdateBy` operators

### LINKED ISSUE #2760: Feature request: Exponential Decaying Sum UpdateBy operation
Very similar to exponential moving average, this keeps track of the sum instead of an average.

### PR DESCRIPTION
**Pull request by lbooker42**
Closes #2760

### CODE REVIEW

File: table-api/src/main/java/io/deephaven/api/updateby/UpdateByOperation.java [hunk 1]
```diff
@@ -214,6 +214,141 @@ static UpdateByOperation Ema(OperationControl control, String timestampColumn, D
...
+     * Create an {@link EmsSpec exponential moving sum} for the supplied column name pairs, using ticks as the decay
+     * unit. Uses the default OperationControl settings.
+     * <p>
+     * The formula used is
+     * </p>
+     *
+     * <pre>
+     *     a = e^(-1 / tickDecay)
+     *     ema_next = a * ema_last + (1 - a) * value
+     * </pre>
```
Comment (devinrsmith) [inline table-api/src/main/java/io/deephaven/api/updateby/UpdateByOperation.java:229]: This should be `ems_` and the formula should be different, right? Ditto w/ other javadoc.

File: java-client/session/src/main/java/io/deephaven/client/impl/UpdateByBuilder.java [hunk 2]
```diff
@@ -118,6 +118,11 @@ public UpdateByColumn.UpdateBySpec visit(EmaSpec ema) {
                     .build();
         }
 
+        @Override
+        public UpdateByColumn.UpdateBySpec visit(EmsSpec spec) {
+            return null;
+        }
```
Comment (devinrsmith) [inline java-client/session/src/main/java/io/deephaven/client/impl/UpdateByBuilder.java:124]: Better to throw unsupported here IMO and link to ticket for gRPC side. I notice the rolling impls are the same, probably best to update those w/ unsupported exceptions as well.

File: java-client/session/src/test/java/io/deephaven/client/impl/UpdateBySpecBuilderTest.java [hunk 3]
```diff
@@ -41,6 +41,11 @@ public UpdateByColumn.UpdateBySpec visit(EmaSpec ema) {
                     .build();
         }
 
+        @Override
+        public UpdateByColumn.UpdateBySpec visit(EmsSpec spec) {
+            return null;
+        }
```
Comment (devinrsmith) [inline java-client/session/src/test/java/io/deephaven/client/impl/UpdateBySpecBuilderTest.java:47]: Impl is fine, but we should link to ticket.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/DoubleEMSOperator.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
+/*
+ * ---------------------------------------------------------------------------------------------------------------------
+ * AUTO-GENERATED CLASS - DO NOT EDIT MANUALLY - for any changes edit CharEMSOperator and regenerate
+ * ---------------------------------------------------------------------------------------------------------------------
+ */
+/*
+ * ---------------------------------------------------------------------------------------------------------------------
+ * AUTO-GENERATED CLASS - DO NOT EDIT MANUALLY - for any changes edit FloatEMSOperator and regenerate
+ * ---------------------------------------------------------------------------------------------------------------------
+ */
```
Comment (devinrsmith) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/DoubleEMSOperator.java:10]: Double headers?

File: engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/BasePrimitiveEMSOperator.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+import io.deephaven.engine.table.impl.util.RowRedirection;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import static io.deephaven.util.QueryConstants.NULL_LONG;
+
+public abstract class BasePrimitiveEMSOperator extends BaseDoubleUpdateByOperator {
+    protected final OperationControl control;
+    protected final double alpha;
+    protected double oneMinusAlpha;
```
Comment (devinrsmith) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/BasePrimitiveEMSOperator.java:19]: Is this used for EMS?

**Review by devinrsmith — COMMENTED** [review]
Quick overview, didn't get deep into the operator logic. Will trial out the operator and report back.

**Comment by devinrsmith** [comment]
Mentioned in DM, but there may be a way to share operator implementations among Exponential Moving Avg / Sum / Min / Max. There also may be ways to share construction of them to make it easier for end user?

For example, it might be cool to have specifically typed helpers off of existing specs:

```python
avg().exponential(window(...))
sum_().exponential(window(...))
avg().rolling(window(...))
...
```

**Comment by lbooker42** [comment]
Creating a shared Exponential Moving implementations is clever and I will refactor in a follow-on PR (likely `Exponential Moving Min/Max`)

File: engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/BaseBigNumberEMSOperator.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        public void reset() {
+            curVal = null;
+            lastStamp = NULL_LONG;
+            lastDt = NULL_LONG;
+            alpha = null;
+        }
+    }
+
+    /**
+     * An operator that computes an EMA from a big number column using an exponential decay function.
```
Comment (nbauernfeind) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/BaseBigNumberEMSOperator.java:59]: s/EMA/EMS

File: engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/BasePrimitiveEMSOperator.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+     * @param pair the {@link MatchPair} that defines the input/output for this operation
+     * @param affectingColumns the names of the columns that affect this ema
+     * @param rowRedirection the row redirection to use for the EMA output columns
+     * @param control the control parameters for EMA
+     * @param timestampColumnName an optional timestamp column. If this is null, it will be assumed time is measured in
+     *        integer ticks.
+     * @param windowScaleUnits the smoothing window for the EMA. If no {@code timestampColumnName} is provided, this is
+     *        measured in ticks, otherwise it is measured in nanoseconds.
+     */
+    public BasePrimitiveEMSOperator(@NotNull final MatchPair pair,
```
Comment (nbauernfeind) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/BasePrimitiveEMSOperator.java:46]: s/EMA/EMS in a few places of the javadoc for this method

File: engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/BigDecimalEMSOperator.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+            // chunk processing
+            if (timestampColumnName == null) {
+                // compute with ticks
+                for (int ii = 0; ii < len; ii++) {
+                    // read the value from the values chunk
+                    final BigDecimal input = objectValueChunk.get(ii);
+                    if (input == null) {
+                        handleBadData(this, true);
+                    } else {
+                        if (curVal == null) {
```
Comment (nbauernfeind) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/BigDecimalEMSOperator.java:41]: I would combine the `else` and `if` to decrease a layer of unnecessary indentation.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/BigDecimalEMSOperator.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+                    final BigDecimal input = objectValueChunk.get(ii);
+                    final long timestamp = tsChunk.get(ii);
+                    final boolean isNull = input == null;
+                    final boolean isNullTime = timestamp == NULL_LONG;
+                    if (isNull) {
+                        handleBadData(this, isNull);
+                    } else if (isNullTime) {
+                        // no change to curVal and lastStamp
+                    } else {
+                        if (curVal == null) {
```
Comment (nbauernfeind) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/BigDecimalEMSOperator.java:64]: Same here on combing `else` and `if`.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/BigDecimalEMSOperator.java [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+                }
+            }
+
+            // chunk output to column
+            writeToOutputColumn(inputKeys);
+        }
+
+        @Override
+        public void push(int pos, int count) {
+            throw new IllegalStateException("EMAOperator#push() is not used");
```
Comment (nbauernfeind) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/BigDecimalEMSOperator.java:90]: The codebase seems to prefer an assertion error instead:
```throw Assert.statementNeverExecuted( ... )```

File: engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/BigIntegerEMSOperator.java [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+                }
+            }
+
+            // chunk output to column
+            writeToOutputColumn(inputKeys);
+        }
+
+        @Override
+        public void push(int pos, int count) {
+            throw new IllegalStateException("EMAOperator#push() is not used");
```
Comment (nbauernfeind) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/BigIntegerEMSOperator.java:94]: ```Assert.statementNeverExecuted```

File: engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/BigIntegerEMSOperator.java [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+     *
+     * @param pair                the {@link MatchPair} that defines the input/output for this operation
+     * @param affectingColumns    the names of the columns that affect this ema
+     * @param rowRedirection      the {@link RowRedirection} to use for dense output sources
+     * @param control             defines how to handle {@code null} input values.
+     * @param timestampColumnName the name of the column containing timestamps for time-based calcuations
+     * @param windowScaleUnits      the smoothing window for the EMA. If no {@code timestampColumnName} is provided, this is measured in ticks, otherwise it is measured in nanoseconds
+     * @param valueSource         a reference to the input column source for this operation
+     */
+    public BigIntegerEMSOperator(@NotNull final MatchPair pair,
```
Comment (nbauernfeind) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/BigIntegerEMSOperator.java:109]: s/EMA/EMS in javadoc

File: engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/CharEMSOperator.java [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+     *
+     * @param pair                the {@link MatchPair} that defines the input/output for this operation
+     * @param affectingColumns    the names of the columns that affect this ema
+     * @param rowRedirection      the {@link RowRedirection} to use for dense output sources
+     * @param control             defines how to handle {@code null} input values.
+     * @param timestampColumnName the name of the column containing timestamps for time-based calcuations
+     * @param windowScaleUnits      the smoothing window for the EMA. If no {@code timestampColumnName} is provided, this is measured in ticks, otherwise it is measured in nanoseconds
+     * @param valueSource         a reference to the input column source for this operation
+     */
+    public CharEMSOperator(@NotNull final MatchPair pair,
```
Comment (nbauernfeind) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/CharEMSOperator.java:118]: s/EMA/EMS in javadoc

File: engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/BigIntegerEMSOperator.java [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+                    } else if (isNullTime) {
+                        // no change to curVal and lastStamp
+                    } else {
+                        final BigDecimal decimalInput = new BigDecimal(input, control.bigValueContextOrDefault());
+                        if (curVal == null) {
+                            curVal = decimalInput;
+                            lastStamp = timestamp;
+                        } else {
+                            final long dt = timestamp - lastStamp;
+                            if (dt != 0) {
```
[the remark this answers is not in the document]
Comment (lbooker42) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emsum/BigIntegerEMSOperator.java:72]: Yes, I realized that ignoring `dt == 0` is completely wrong for most exponential moving operations.

**Review by devinrsmith — APPROVED** [review]

**Review by nbauernfeind — APPROVED** [review]

**Comment by deephaven-internal** [comment]
Labels indicate documentation is required. Issues for documentation have been opened:

How-to: https://github.com/deephaven/deephaven.io/issues/2452
Conceptual: https://github.com/deephaven/deephaven.io/issues/2450
Reference: https://github.com/deephaven/deephaven.io/issues/2451
Blog: https://github.com/deephaven/deephaven.io/issues/2453

**Pull request merged**